src/TopDownProteomics/ProForma/ChemicalProteoformHashGenerator.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using TopDownProteomics.Chemistry;
+using TopDownProteomics.Proteomics;
+
+namespace TopDownProteomics.ProForma
+{
+    /// <summary>Creates a chemical proteoform hash from a proteoform group.</summary>
+    public class ChemicalProteoformHashGenerator
+    {
+        /// <summary>Generates a chemical proteoform hash for the specified proteoform group.</summary>
+        /// <param name="proteoformGroup">The proteoform group.</param>
+        /// <returns></returns>
+        public string Generate(IProteoformGroup proteoformGroup)
+        {
+            ProFormaDescriptor nTermDescriptor =
+                this.GetFormulaDescriptor(proteoformGroup.NTerminalModification);
+            IList<ProFormaDescriptor> nTermDescriptors = nTermDescriptor == null ? null : new[] { nTermDescriptor };
+
+            ProFormaDescriptor cTermDescriptor =
+                this.GetFormulaDescriptor(proteoformGroup.CTerminalModification);
+            IList<ProFormaDescriptor> cTermDescriptors = cTermDescriptor == null ? null : new[] { cTermDescriptor };
+
+            IList<ProFormaTag> tags = new List<ProFormaTag>();
+            if (proteoformGroup.Modifications != null)
+            {
+                foreach (IProteoformModificationWithIndex proteoformModificationWithIndex in proteoformGroup.Modifications)
+                {
+                    ProFormaDescriptor descriptor = this.GetFormulaDescriptor(proteoformModificationWithIndex);
+                    tags.Add(new ProFormaTag(proteoformModificationWithIndex.ZeroBasedIndex, new[] { descriptor }));
+                }
+            }
+
+            ProFormaTerm proFormaTerm = new ProFormaTerm(proteoformGroup.GetSequence(), null, nTermDescriptors, cTermDescriptors, tags.OrderBy(t => t.Descriptors.First().Value).ToArray());
```

Comment:
_Random thought_: I want to keep an eye on these next 3 lines. At first I don't like them because it doesn't make things composeable, e.g. it might be nice to simply return the `ProFormaTerm` and allow the client to operate on that, add another public method to write the string, etc. But at this point, I think I like hiding everything behind this one public method.